src/ObjectUtils/LinuxMap.cpp

```diff
@@ -89,28 +89,156 @@ ErrorMessageOr<std::vector<ModuleInfo>> ReadModules(int32_t pid) {
+        map_end >= base_address_ + cached_coff_file_->GetExecutableSegmentOffset() +
+                       cached_coff_file_->GetExecutableSegmentSize()) {
+      ORBIT_LOG("Guessing that executable map at %#x-%#x belongs to \"%s\"", map_start, map_end,
+                file_path_);
+      MarkExecutableMapEncountered();
+      return true;
+    }
+
+    ORBIT_LOG("%s", error_message);
+    coff_text_section_map_could_be_encountered_ = false;
+    cached_coff_file_ = nullptr;
+    return false;
+  }
+
+ private:
+  std::string file_path_;
+  // The address at which the first byte of the file is (or would be) mapped.
+  uint64_t base_address_;
+  // False if not a PE, if the .text segment has already been found, or if we are already past the
+  // address at which we could find the .text segment.
+  bool coff_text_section_map_could_be_encountered_ = true;
+  std::unique_ptr<ObjectFile> cached_coff_file_;
+};
+}  // namespace
+
 ErrorMessageOr<std::vector<ModuleInfo>> ParseMaps(std::string_view proc_maps_data) {
   const std::vector<std::string> proc_maps = absl::StrSplit(proc_maps_data, '\n');
 
   std::vector<ModuleInfo> result;
 
+  // This is used to detect mappings that correspond to the .text section of a PE but that are not
+  // file-backed because the file alignment doesn't satisfy the requirements of mmap.
+  std::optional<FileMappedIntoMemory> last_file_mapped_into_memory;
+
   for (const std::string& line : proc_maps) {
     std::vector<std::string> tokens = absl::StrSplit(line, ' ', absl::SkipEmpty());
-    // tokens[4] is the inode column. If inode equals 0, then the memory is not
-    // mapped to a file (might be heap, stack or something else)
-    if (tokens.size() != 6 || tokens[4] == "0") continue;
-
-    const std::string& module_path = tokens[5];
+    if (tokens.size() < 5) continue;
+    // tokens[4] is the inode column. If inode equals 0, then the memory is not backed by a file.
+    // If a map not backed by a file has a name, it's a special one like [stack], [heap], etc.
+    if (tokens[4] == "0" && tokens.size() == 6) continue;
 
     std::vector<std::string> addresses = absl::StrSplit(tokens[0], '-');
     if (addresses.size() != 2) continue;
 
     uint64_t start = std::stoull(addresses[0], nullptr, 16);
     uint64_t end = std::stoull(addresses[1], nullptr, 16);
-    bool is_executable = tokens[1].size() == 4 && tokens[1][2] == 'x';
+    uint64_t offset = std::stoull(tokens[2], nullptr, 16);
+
+    std::string module_path;
+    if (tokens[4] != "0") {  // The mapping is file-backed.
+      if (tokens.size() == 6) {
+        module_path = tokens[5];
+        if (!last_file_mapped_into_memory.has_value() ||
+            module_path != last_file_mapped_into_memory->GetFilePath()) {
+          last_file_mapped_into_memory.emplace(module_path, start, offset);
+        }
+      } else {  // Unexpected: the mapping is file-backed but no path is present.
+        last_file_mapped_into_memory.reset();
```

Comment:
Maybe add a log for that case?

Reply:
Done. This actually revealed http://b/228826044 which is to be tackled before this change and independently by https://github.com/google/orbit/pull/3571. This change will need to be adjusted once that one is in.